File: game-engine/src/main/java/org/luxons/sevenwonders/game/moves/BuildWonderMove.java
package org.luxons.sevenwonders.game.moves;

import java.util.List;

import org.luxons.sevenwonders.game.Settings;
import org.luxons.sevenwonders.game.api.PlayerMove;
import org.luxons.sevenwonders.game.api.Table;
import org.luxons.sevenwonders.game.boards.Board;
import org.luxons.sevenwonders.game.cards.Card;

public class BuildWonderMove extends CardFromHandMove {

    BuildWonderMove(int playerIndex, Card card, PlayerMove move) {
        super(playerIndex, card, move);
    }

    @Override
    public void validate(Table table, List<Card> playerHand) throws InvalidMoveException {
        super.validate(table, playerHand);
        Board board = table.getBoard(getPlayerIndex());
        if (!board.getWonder().isNextStageBuildable(table, getPlayerIndex(), getTransactions())) {
            throw new InvalidMoveException(
                    String.format("Player %d cannot upgrade his wonder with the given resources", getPlayerIndex()));
        }
    }

    @Override
    public void place(Table table, List<Card> discardedCards, Settings settings) {
        Board board = table.getBoard(getPlayerIndex());
        board.getWonder().buildLevel(getCard().getBack());
    }

    @Override
    public void activate(Table table, List<Card> discardedCards, Settings settings) {
        int playerIndex = getPlayerIndex();
        Board board = table.getBoard(playerIndex);
        board.getWonder().activateLastBuiltStage(table, playerIndex, getTransactions());
    }
}

File: game-engine/src/main/kotlin/org/luxons/sevenwonders/game/moves/BuildWonderMove.kt
package org.luxons.sevenwonders.game.moves

import org.luxons.sevenwonders.game.PlayerContext
import org.luxons.sevenwonders.game.Settings
import org.luxons.sevenwonders.game.api.PlayerMove
import org.luxons.sevenwonders.game.cards.Card

internal class BuildWonderMove(move: PlayerMove, card: Card, player: PlayerContext) :
    CardFromHandMove(move, card, player) {

    private val wonder = player.board.wonder

    init {
        if (!wonder.isNextStageBuildable(playerContext.board, transactions)) {
            throw InvalidMoveException(this, "all levels are already built, or the given resources are insufficient")
        }
    }

    override fun place(discardedCards: MutableList<Card>, settings: Settings) = wonder.placeCard(card.back)

    override fun activate(discardedCards: List<Card>, settings: Settings) =
        wonder.activateLastBuiltStage(playerContext, transactions)
}
